ARCs/arc-0072.md

```diff
@@ -0,0 +1,342 @@
+      ],
+      "returns": { "type": "void" }
+    },
+    {
+      "name": "setApprovalForAll",
+      "desc": "Approve an operator for all NFTs for a user",
+      "readonly": false,
+      "args": [
+        { "type": "address", "name": "operator", "desc": "Approved operator address" },
+        { "type": "bool", "name": "approved", "desc": "true to give approval, false to revoke" },
+      ],
+      "returns": { "type": "void" }
+    },
+    {
+      "name": "getApproved",
+      "desc": "Get the current approved address for a single NFT",
+      "readonly": true,
+      "args": [
+        { "type": "uint256", "name": "tokenId", "desc": "The ID of the NFT" },
+      ],
+      "returns": { "type": "address", "desc": "address of approved user or zero" }
+    },
+    {
+      "name": "isApprovedForAll",
+      "desc": "Query if an address is an authorized operator for another address",
+      "readonly": true,
+      "args": [
+        { "type": "address", "name": "owner" },
+        { "type": "address", "name": "operator" },
+      ],
+      "returns": { "type": "bool", "desc": "whether operator is authorized for all NFTs of owner" }
+    },
+  ],
+  "events": [
+    {
+      "name": "Approval",
+      "desc": "An address has been approved to transfer ownership of the NFT",
+      "args": [
+        {
+          "type": "address",
+          "name": "owner",
+          "desc": "The current owner of the NFT"
+        },
+        {
+          "type": "address",
+          "name": "approved",
+          "desc": "The approved user for the NFT"
+        },
+        {
+          "type": "uint256",
+          "name": "tokenId",
+          "desc": "The ID of the NFT"
+        }
+      ]
+    },
+    {
+      "name": "ApprovalForAll",
+      "desc": "Operator set or unset for all NFTs defined by this contract for an owner",
+      "args": [
+        {
+          "type": "address",
+          "name": "owner",
+          "desc": "The current owner of the NFT"
+        },
+        {
+          "type": "address",
+          "name": "operator",
+          "desc": "The approved user for the NFT"
+        },
+        {
+          "type": "bool",
+          "name": "approved",
+          "desc": "The ID of the NFT"
+        }
+      ]
+    },
+  ]
+}
+```
+
+The `Approval` event MUST be emitted when the `approve` method is called successfully.
+The zero address for the `approve` method and the `Approval` event indicate no approval, including revocation of previous single NFT controller.
+When a `Transfer` event emits, this also indicates that the approved address for that NFT (if any) is reset to none.
+The `ApprovalForAll` event MUST be emitted when the `setApprovalForAll` method is called successfully.
+The contract MUST allow multiple operators per owner.
+The `transferFrom` method, when its `nftId` argument is owned by its `from` argument, MUST succeed for when called by an address that is approved for the given NFT or approved as operator for the owner.
+
```

Comment:
This section needs to specify under what conditions an operator can be set. For example, the owner has to be the one calling. But also, can an operator for a single NFT designate a new operator? Can an operator for a user add more operators?

Reply:
We are only designing the ABI and not the logic behind a specific NFT-Application

Reply:
I agree we need to be careful about getting too specific. However, if a contract implemented this spec as-is, then I could set myself as the operator for every NFT because there is nothing in the spec requiring authorization from current owners and/or operators. So I think something should be added into the spec to prevent that.